compiler/vm/vmdef.nim

```diff
@@ -519,6 +580,143 @@ type
                             ## is active
 
     flags*: set[CodeGenFlag] ## input
+  
+  VmGenDiagKind* = enum
+    vmGenDiagMissingImportcCompleteStruct
+    vmGenDiagTooManyRegistersRequired
+    vmGenDiagCannotFindBreakTarget
+    vmGenDiagNotUnused
+    vmGenDiagNotAFieldSymbol
+    vmGenDiagTooLargeOffset
+    vmGenDiagCannotGenerateCode
+    vmGenDiagCannotCast
+    vmGenDiagBadExpandToAst
+    vmGenDiagCannotEvaluateAtComptime
+    vmGenDiagCannotImportc
+    vmGenDiagInvalidObjectConstructor
+    vmGenDiagNoClosureIterators
+    vmGenDiagCannotCallMethod
+    vmGenDiagNotAField
+  
+  VmTypeMismatch* = object
+    actualType*, formalType*: PType
+
+  VmGenDiag* = object
+    ## `Diag`nostic data from VM Gen, mostly errors
+    ast*: PNode
+    typ*: PType
+    msg*: string
+    sym*: PSym
+    location*: TLineInfo        ## diagnostic location
+    instLoc*: InstantiationInfo ## instantiation in VM Gen's source
+    case kind*: VmGenDiagKind
+      of vmGenDiagCannotCast:
+        typeMismatch*: VmTypeMismatch
+      else:
+        discard
+
+  GenDiagHandler* = proc(c: TCtx, diag: VmGenDiag): void
+  
+  VmEventKind* = enum
+    # errors
+    vmEvtOpcParseExpectedExpression
+    vmEvtUserError
+    vmEvtUnhandledException
+    vmEvtCannotCast
+    vmEvtCallingNonRoutine
+    vmEvtCannotModifyTypechecked
+    vmEvtNilAccess
+    vmEvtAccessOutOfBounds
+    vmEvtAccessTypeMismatch
+    vmEvtAccessNoLocation
+    vmEvtErrInternal
+    vmEvtIndexError
+    vmEvtOutOfRange
+    vmEvtOverOrUnderflow
+    vmEvtDivisionByConstZero
+    vmEvtNodeNotASymbol
+    vmEvtNodeNotAProcSymbol
+    vmEvtIllegalConv
+    vmEvtMissingCacheKey
+    vmEvtCacheKeyAlreadyExists
+    vmEvtFieldNotFound
+    vmEvtNotAField
+    vmEvtFieldInavailable
+    vmEvtCannotSetChild
+    vmEvtCannotAddChild
+    vmEvtCannotGetChild
+    vmEvtNoType
+    vmEvtUnsupportedNonNil
+    vmEvtTooManyIterations
+    vmEvtQuit
+    # debug / trace
+    vmEvtDebugCodeListing
+    vmEvtDebugExecTraceMinimal
+    vmEvtDebugExecTraceFull
+
+  VmEventKindAccessError* = range[vmEvtAccessOutOfBounds .. vmEvtAccessNoLocation]
+
+  VmDebugCodeEntry* = object
+    isTarget*: bool
+    info*: TLineInfo
+    pc*: int
+    idx*: int
+    case opc*: TOpcode:
+      of opcConv, opcCast:
+        types*: tuple[tfrom, tto: PType]
+      of opcLdConst, opcAsgnConst:
+        ast*: PNode
+      else:
+        discard
+    ra*: int
+    rb*: int
+    rc*: int
+
+  VmEvent* = object
+    ## Event data from a VM instance, mostly errors
+    ast*: PNode
```

Comment:
Should be able to move these fields into the appropriate branches now that we don't have a super object.